### Part 1. NumPy NDArrays as is.

In [1]:
import numpy as np

In [2]:
arr = np.array([1,2, 3, 4])

In [3]:
arr

array([1, 2, 3, 4])

In [6]:
arr2 = np.array([1, 2, 3, 4], dtype=np.float)

In [7]:
arr2

array([1., 2., 3., 4.])

In [8]:
arr2.dtype

dtype('float64')

In [9]:
arr3 = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]], dtype=np.float)

In [10]:
arr3

array([[1., 2., 3.],
       [4., 5., 6.],
       [7., 8., 9.]])

In [11]:
np.shape(arr3)

(3, 3)

In [12]:
arr3.shape

(3, 3)

### Part 2. NumPy NDArray from Pandas.

In [13]:
import pandas as pd

In [15]:
df = pd.DataFrame()
x = df.values

In [16]:
type(x)

numpy.ndarray

In [17]:
my_series = pd.Series([5, 6, 7, 8, 9, 10])

In [18]:
my_series.values

array([ 5,  6,  7,  8,  9, 10])

### Part 3. Another ways creating NumPy NDArrays.

#### одномерный массив из пяти элементов, память для которого выделена, но не инициализирована

In [19]:
np.empty(5)

array([4.65236272e-310, 0.00000000e+000, 4.65236286e-310, 4.65236279e-310,
       4.65236279e-310])

#### массив размером 10x7, заполненный нулями

In [20]:
np.zeros((10, 7))

array([[0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.]])

#### массив размером 3х3х3, заполненный единицами

In [21]:
np.ones((3, 3, 3))

array([[[1., 1., 1.],
        [1., 1., 1.],
        [1., 1., 1.]],

       [[1., 1., 1.],
        [1., 1., 1.],
        [1., 1., 1.]],

       [[1., 1., 1.],
        [1., 1., 1.],
        [1., 1., 1.]]])

#### единичная матрица (элементы главной диагонали равны 1, остальные — 0) размера 3х3

In [22]:
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

#### массив 3x5 заполненный числом 3.14

In [24]:
np.full((3, 5), 3.14)

array([[3.14, 3.14, 3.14, 3.14, 3.14],
       [3.14, 3.14, 3.14, 3.14, 3.14],
       [3.14, 3.14, 3.14, 3.14, 3.14]])

#### одномерный массив, заполненный числами в диапазоне от 0 до 20 с шагом 7

In [25]:
np.arange(0, 21, 7)

array([ 0,  7, 14])

#### массив из пяти чисел, равномерно распределённых в интервале между 0 и 1 включительно

In [26]:
np.linspace(0, 1, 5)

array([0.  , 0.25, 0.5 , 0.75, 1.  ])

#### массив размера 3х3, заполненный случайными числами из диапазона от 0 до 10

In [27]:
np.random.randint(0, 10, (3, 3))

array([[7, 0, 1],
       [0, 7, 8],
       [6, 0, 8]])

### Part 4.Tasks.

#### Task 1.
Какой индекс имеет верхний левый элемент любого двумерного массива?

#### Solution:

In [28]:
arr4 = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
arr4[0, 0]

1

#### Conclusion: [0, 0] is the right answer.

#### Task 2.
``` python
np.array([[8,13,18,4], [25,3,18,28], [14,15,18,24], [1,2,3,4]]) 
```
Какие элементы расположены на главной диагонали массива, заданного с помощью кода выше?
* 8, 3, 18, 4
* 18, 18, 18, 1
* 25, 3, 18, 24
* 4, 28, 24, 4
* этот массив не имеет главной диагонали 

####  Solution:

In [30]:
np.array([[8,13,18,4], [25,3,18,28], [14,15,18,24], [1,2,3,4]])

array([[ 8, 13, 18,  4],
       [25,  3, 18, 28],
       [14, 15, 18, 24],
       [ 1,  2,  3,  4]])

#### Conclusion: 8, 3, 18, 4 is the right answer!!!

#### Task 3.
``` python
my_secret = [x for x in range(1, 301, 7) if x%10 == 7 or x%10 == 1]
np.array([my_secret, [x/2 for x in my_secret], [x-100 for x in my_secret]])
```
Какие элементы расположены в последнем столбце массива, заданного с помощью кода выше?
* 301, 150.5, 201
* 301.0, 150.5, 201.0
* -99.0, 0.5, 1.0
* 281.0, 140.5, 181.0
* нет правильного ответа 

#### Solution:

In [31]:
my_secret = [x for x in range(1, 301, 7) if x%10 == 7 or x%10 == 1]
np.array([my_secret, [x/2 for x in my_secret], [x-100 for x in my_secret]])

array([[  1. ,  57. ,  71. , 127. , 141. , 197. , 211. , 267. , 281. ],
       [  0.5,  28.5,  35.5,  63.5,  70.5,  98.5, 105.5, 133.5, 140.5],
       [-99. , -43. , -29. ,  27. ,  41. ,  97. , 111. , 167. , 181. ]])

#### Conclusion: 281, 140.5, 181 is the right answer!!!

#### Task 4.
С помощью каких вариантов кода можно создать массив, среднее арифметическое всех элементов главной диагонали которого равно 1?
* ``` python
np.ones(5)
```
* ``` python
np.ones(5,5)
```
* ``` python
np.ones((5,5))
```
* ``` python
np.eye(5)
```
* ``` python
np.full((5,5), 1)
```
* ``` python
np.array([[1,2,3,4,5], [5,1,2,3,4], [4,5,1,2,3], [3,4,5,1,2], [2,3,4,5,1]]) 
```

#### Solution:

In [32]:
np.ones(5)

array([1., 1., 1., 1., 1.])

In [33]:
np.ones(5, 5)

TypeError: data type not understood

In [40]:
# 1st way
np.ones((5, 5))

array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])

In [41]:
# 2nd way
np.eye(5)

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

In [42]:
# 3rd way
np.full((5, 5), 1)

array([[1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1]])

In [43]:
# 4th way
np.array([[1,2,3,4,5], [5,1,2,3,4], [4,5,1,2,3], [3,4,5,1,2], [2,3,4,5,1]])

array([[1, 2, 3, 4, 5],
       [5, 1, 2, 3, 4],
       [4, 5, 1, 2, 3],
       [3, 4, 5, 1, 2],
       [2, 3, 4, 5, 1]])

#### Conclusion: 3rd, 4th, 5th and 6th ways are right aswers!!!

#### Task 5.
``` python
[[1, 1, 1],
 [0, 2, 2],
 [2, 4, 1]]
```
С помощью каких вариантов кода НЕВОЗМОЖНО создать массив, представленный в примере выше?
* ``` python
np.random.randint(0, 4, (3, 3))
```
* ``` python
np.random.randint(0, 5, (3, 3))
```
* ``` python
np.random.randint(0, 6, (3, 3))
```
* ``` python
np.random.randint(0, 5, 3) 
```

#### Solution:
1st way can't generate 4 and 4th way generate 1D array.

#### Conclusion: 1st and 4th ways are right answers!!!

#### Task 6.
Какие из представленных вариантов кода можно использовать для создания массива 4х4, полностью заполненного единицами?
* ``` python
np.array([[1, 1, 1, 1], [1, 1, 1, 1], [1, 1, 1, 1], [1, 1, 1, 1]])
```
* ``` python
np.eye(4)
```
* ``` python
np.ones((4, 4), dtype=float)
```
* ``` python
np.full((4, 4), 1) 
```

#### Solution:
2nd way generates eigen array, i.e. ones will be onlny on main diagonal amd onther elements will be zeros.

#### Conclusion: 1st, 3rd and 4th ways are right answers!!!

#### Task 7.
Какие из вариантов кода можно использовать для создания массива, заполненного нечётными числами?
* ``` python
np.array(range(9))
```
* ``` python
np.ones((2,2,2))
```
* ``` python
np.array([i for i in range(11) if i%2])
```
* ``` python
np.arange(1, 14, 2) 
```

#### Solution:
1st way generates both odd and even numbers array.

#### Conclusion: 2nd, 3rd and 4th ways are right answers!!!